lib/Dialect/FIRRTL/Transforms/LegacyWiring.cpp

```diff
@@ -0,0 +1,120 @@
+//===- LegacyWiring- legacy Wiring annotation resolver ----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the legacy Wiring annotation resolver.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Support/LogicalResult.h"
+
+using namespace circt;
+using namespace firrtl;
+
+/// Consume SourceAnnotation and SinkAnnotation, storing into state
+LogicalResult circt::firrtl::applyWiring(const AnnoPathValue &target,
+                                         DictionaryAttr anno,
+                                         ApplyState &state) {
+  auto *context = anno.getContext();
+  ImplicitLocOpBuilder builder(target.ref.getOp()->getLoc(), context);
+
+  // Convert target to Value
+  SmallVector<Value> targetsValues;
+  if (auto portTarget = target.ref.dyn_cast<PortAnnoTarget>()) {
+    auto portNum = portTarget.getImpl().getPortNo();
+    if (auto module = dyn_cast<FModuleOp>(portTarget.getOp())) {
+      builder.setInsertionPointToEnd(module.getBodyBlock());
+      targetsValues.push_back(getValueByFieldID(
```

Comment:
So, IMO, this is something we need to solve more robustly re:WiringProblem infra improvements, but for now it's the responsibility of the problem creation code to ensure a number of things that are a little tricky.  Hopefully we can consolidate/improve things in the future.  Anyway.

Here, inserting into the end of the block makes sense if this is a sink, but for the source you probably want the value to dominate everything the target port dominates (so it can be connected to the same sorts of things, for example down through an instance).

Reply:
Is it enough to insert before/ at the start for sources?